libs/s25main/figures/nofMiner.cpp

```diff
@@ -71,20 +76,100 @@ helpers::OptionalEnum<GoodType> nofMiner::ProduceWare()
+        return FindPointWithResource(GetRequiredResType(), MINER_RADIUS_SETTLERSIV).isValid();
+    else
+        return FindPointWithResource(GetRequiredResType()).isValid();
 }
 
-bool nofMiner::StartWorking()
+MiningBehavior nofMiner::GetMiningBehavior() const
 {
-    MapPoint resPt = FindPointWithResource(GetRequiredResType());
-    if(!resPt.isValid())
-        return false;
     const GlobalGameSettings& settings = world->GetGGS();
-    bool inexhaustibleRes = settings.isEnabled(AddonId::INEXHAUSTIBLE_MINES)
-                            || (workplace->GetBuildingType() == BuildingType::GraniteMine
-                                && settings.isEnabled(AddonId::INEXHAUSTIBLE_GRANITEMINES));
-    if(!inexhaustibleRes)
-        world->ReduceResource(resPt);
+
+    switch(workplace->GetBuildingType())
+    {
+        case BuildingType::GoldMine: return (MiningBehavior)settings.getSelection(AddonId::MINING_OVERHAUL_GOLD);
+        case BuildingType::IronMine: return (MiningBehavior)settings.getSelection(AddonId::MINING_OVERHAUL_IRON);
+        case BuildingType::CoalMine: return (MiningBehavior)settings.getSelection(AddonId::MINING_OVERHAUL_COAL);
+        case BuildingType::GraniteMine: return (MiningBehavior)settings.getSelection(AddonId::MINING_OVERHAUL_GRANITE);
+        default: return MiningBehavior::Normal;
+    }
+}
+
+bool nofMiner::StartWorking()
+{
+    isAlteredWorkcycle = false;
+
+    MiningBehavior addonSettings = GetMiningBehavior();
+
+    switch (addonSettings)
+    {
+        case MiningBehavior::S4Like:
+        {
+            int sumResAmount = 0;
+
+            MapPoint nonMinumResPt;
+
+            const std::vector<MapPoint> reachablePts =
+              world->GetPointsInRadiusWithCenter(workplace->GetPos(), MINER_RADIUS_SETTLERSIV);
+
+            for(const MapPoint curPt : reachablePts)
+            {
+                const Resource resPt = world->GetNode(curPt).resources;
+
+                if(resPt.getType() != GetRequiredResType())
+                    continue;
+
+                const auto resAmount = resPt.getAmount();
+
+                sumResAmount += resAmount;
+
+                if(resAmount > 1u && !nonMinumResPt.isValid())
+                    nonMinumResPt = curPt;
+            }
+
+            // depending on remaining resources, roll if this workcycle needs to be altered or not
+            if(RANDOM_RAND(world->numPointsRadius(MINER_RADIUS_SETTLERSIV, true) * MINER_MAX_QUANTITY)
+               > sumResAmount)
+            {
+                isAlteredWorkcycle = true;
+            } else
+            {
+                if(nonMinumResPt.isValid())
+                    world->ReduceResource(nonMinumResPt);
+            }
+        }
+        break;
+        // TODO: Is this required? FindPointWithResource checks if there is at least one
+        case MiningBehavior::Inexhaustible:
+        {
+            const MapPoint resPt = FindPointWithResource(GetRequiredResType());
+            if(!resPt.isValid())
```

Comment:
See previous change:
```suggestion
            if(!FindPointWithResource(GetRequiredResType()).isValid())
```